File: cabot_influxdb_client/script/diagnostics_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from datetime import datetime
from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus
from influxdb_client import InfluxDBClient, Point, WriteOptions
from influxdb_client.client.write_api import SYNCHRONOUS

from rosidl_runtime_py import message_to_yaml

USERNAME = "cabot"
PASSWORD = "cabot-influxdb"
TOKEN = "a54a87f7-73a0-4534-9741-ad7ff4e7d111"
ORGANIZATION_NAME = "cabot"
BUCKET_NAME = "grafana-test"
host = "localhost"
port = 8086

influxdb_url = f"http://{host}:{port}"
token = TOKEN
org = ORGANIZATION_NAME
bucket = BUCKET_NAME

client = InfluxDBClient(url=influxdb_url, token=token, org=org)
write_api = client.write_api(write_options=SYNCHRONOUS)


def getLevelText(level):
    if level == DiagnosticStatus.OK:
        return "OK"
    elif level == DiagnosticStatus.WARN:
        return "WARN"
    elif level == DiagnosticStatus.ERROR:
        return "ERROR"
    elif level == DiagnosticStatus.STALE:
        return "STALE"
    else:
        return "UNKNOWN"


class DiagnosticsListener(Node):
    def __init__(self):
        super().__init__('pose_log_listener')
        self.subscription = self.create_subscription(DiagnosticArray, '/diagnostics_agg', self.diagnostics_callback, 10)
        self.subscription2 = self.create_subscription(DiagnosticStatus, '/diagnostics_toplevel_state', self.diagnostics2_callback, 10)
        self.client = InfluxDBClient(url='http://localhost:8086', token=token, org=org)
        self.write_api = self.client.write_api(write_options=SYNCHRONOUS)

    def diagnostics_callback(self, msg):
        target_name = "Soft: Navigation2"
        max_level = 0
        for state in msg.status:
            items = state.name.split("/")
            if len(items) < 3:
                continue
            if items[2] == target_name:
                level = int.from_bytes(state.level, byteorder='big')
                if max_level < level:
                    max_level = level
        level_text = getLevelText(max_level)
        point = Point("diagnostic")\
            .tag("name", target_name)\
            .field("level", max_level)\
            .field("level_text", level_text)\
            .time(datetime.utcnow())
        self.write_api.write(bucket=bucket, org=org, record=point)

    def diagnostics2_callback(self, msg):
        pass
        #print(message_to_yaml(msg))
        #point = Point("pose_data").field("lat", msg.lat).field("lng", msg.lng).field("floor", msg.floor).time(datetime.utcnow())
        #self.write_api.write(bucket=bucket, org=org, record=point)

    def __del__(self):
        self.client.close()

def main(args=None):
    rclpy.init(args=args)
    pose_log_listener = DiagnosticsListener()
    rclpy.spin(pose_log_listener)
    pose_log_listener.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: cabot_influxdb_client/src/diagnostics_client.cpp
#include "diagnostics_client.hpp"

const std::string USERNAME = "cabot";
const std::string PASSWORD = "cabot-influxdb";
const std::string TOKEN = "";
const std::string ORGANIZATION_NAME = "cabot";
const std::string BUCKET_NAME = "grafana-test";
const std::string INFLUXDB_URL = "http://localhost:8086";

std::string getLevelText(int level){
  switch(level){
    case diagnostic_msgs::msg::DiagnosticStatus::OK:
      return "OK";
    case diagnostic_msgs::msg::DiagnosticStatus::WARN:
      return "WARN";
    case diagnostic_msgs::msg::DiagnosticStatus::ERROR:
      return "ERROR";
    case diagnostic_msgs::msg::DiagnosticStatus::STALE:
      return "STALE";
    default:
      return "UNKNOWN";
  }
}

DiagnosticsListener::DiagnosticsListener() : Node("diagnostics_listener"){
  subscription_ = create_subscription<diagnostic_msgs::msg::DiagnosticArray>(
      "/diagnostics_agg", 10, std::bind(&DiagnosticsListener::diagnostics_callback, this, std::placeholders::_1));
  subscription2_ = create_subscription<diagnostic_msgs::msg::DiagnosticStatus>(
      "/diagnostics_toplevel_state", 10, std::bind(&DiagnosticsListener::diagnostics2_callback, this, std::placeholders::_1));
  RCLCPP_INFO(this->get_logger(), "Node is ready");
}

void DiagnosticsListener::diagnostics_callback(const diagnostic_msgs::msg::DiagnosticArray::SharedPtr msg){
  const std::string target_name = "Soft: Navigation2";
  int max_level = 0;
  for(const diagnostic_msgs::msg::DiagnosticStatus& state : msg->status){
    std::vector<std::string> items;
    std::stringstream ss(state.name);
    std::string item;
    while(std::getline(ss, item, '/')){
      items.push_back(item);
    }
    if(items.size() < 3){
      continue;
    }
    if(items[2] == target_name){
      int level = static_cast<int>(state.level);
      max_level = std::max(max_level, level);
    }
  }

  const std::string level_text = getLevelText(max_level);
  influxdb::Point point("diagnostic");
  point.addTag("name", target_name);
  point.addField("level", max_level);
  point.addField("level_text", level_text);
  point.setTimestamp(std::chrono::system_clock::now());

  try{
    std::unique_ptr<influxdb::InfluxDB> idb(influxdb::InfluxDBFactory::Get(INFLUXDB_URL));
    idb->write(std::move(point));
  }catch(const std::exception& e){
    RCLCPP_ERROR(this->get_logger(), "Error writing to InfluxDB: %s", e.what());
  }
}

void DiagnosticsListener::diagnostics2_callback(const diagnostic_msgs::msg::DiagnosticStatus::SharedPtr msg){
  // unimplemented
}

int main(int argc, char* argv[]){
  rclcpp::init(argc, argv);
  std::unique_ptr<DiagnosticsListener> node_ptr(std::make_unique<DiagnosticsListener>());
  std::shared_ptr<DiagnosticsListener> node(std::move(node_ptr));
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
